### ISSUE TITLE
Implement sandbox to run WAL redo routines in

### ISSUE THREAD

**Issue by hlinnaka** [libzenith/postgres#10]

**Comment by knizhnik** [comment]
I wonder what kind of attacks we want to address here?
What kind of vulnerabilities and injections we are expecting here?

Do we assume that Postgres core (without user extension) always produce valid WAL records?

If we are talking about extension, then do we plan to support only well known (verified) extension or any user defined extension? How can extension produce incorrect WAL records? Does extension use custom WAL records (or logical WAL records) and we expect that Postgres is not checking their correctness? Or do we expect that "Byzantine" extension will get direct access to WAL files and intentionally corrupt them? If so, how we are going to protect ourselves from DooS kind of attacks, when WAL records are formally valid, but cause some very expensive operations consuming a lot of disk and CPU resources? Sandbox will not help much in this case, doesn't it?

Final question: if we rewrite redo handlers in some safe language, like Rust, then sandbox will not needed, will it? I looked at redo handlers and it seems to be non-trivial, because them are using a lot AM specific code. But it should be more efficient that running Postgres in VM.

As far as I understand the primary idea of multi-tenant storage   nodes was to reduce per-client overhead. But if we have to proceed WAL records of each client in separate VM, will it be even more expensive than avoid multi-tenancy support at storage level as well as at computation nodes, and running separate Pager process for each client?

**Comment by kelvich** [comment]
> Do we assume that Postgres core (without user extension) always produce valid WAL records?

User can buy exploit for yet-not-disclosed CVE and take over the processing node -- after that we can expect malicious WAL to come into the storage node. It may just crash it or take over the node if there is some bug in redo. Knowing the code that we are running at processing node (postgres redo) this user may look for wrong boundary checks around wal reading and just execute his own code. So IMO the whole attack is not a theoretical one -- attacker will just need to wait few month until next CVE to write malicious WAL and finding exploitable problem in the redo code should also not be a big problem.

> if we rewrite redo handlers in some safe language, like Rust, then sandbox will not needed, will it?
Rust safety means that type system when it is properly used prevents data races and dangling pointers. But array boundary checks are still on programmer -- and that can lead to the problems.

>  it should be more efficient that running Postgres in VM
I believe we may have same performance between host page server <-> postgres in vm as it is between two processes talking to each other via shared memory queue. We actually may avoid running OS in vm at all (not sure that we should, but we can) -- just some parts of postgres reading one memory queue and writing to another one, no files, no processes, etc.

Some links:
https://mirage.io/
https://firecracker-microvm.github.io/
https://qemu.readthedocs.io/en/latest/system/ivshmem.html

**Comment by knizhnik** [comment]
> But array boundary checks are still on programmer -- and that can lead to the problems.
I am not Rust expert, but it seems to be not true.
Rust performs out-of-bounds checking at runtime.
May be there is some compiler switch which allows to disable them or may be compiler can avoid them when it can prove that there is no out-of-bounds access, but by default them are performed:
https://users.rust-lang.org/t/bound-checking-really-present/21809
https://users.rust-lang.org/t/how-to-avoid-bounds-checking/4433/4

**Comment by kelvich** [comment]
Yes, it is usually switched on in a debug build and switched off for release build. We may keep that for release, but working with raw memory means unsafe { } code blocks with all guarantees off. To be able to write something to rust array/vector from WAL you would need to first read len from raw buffer, then memcpy() some data of that len. After that rust runtime can do boundary checks. But will you believe this len in a first place? -- that is what I meant by manual boundary checks. So I don't understand how rewriting redo can generally help. It may be harder or impossible to penetrate such a system, but causing crash/oom can be done instead.

**Comment by kelvich** [comment]
Thinking about that a bit more it seems that we may skip calling VM redo code in getPage requests and use redo only to do periodical WAL chains compaction. So on getPage request we may return page and few records on top of it and use VM only to full page image from time to time (to avoid sending that from compute node) when WAL chain becomes big.

**Comment by adsharma** [comment]
`XLogInsert(RmgrId rmid, uint8 info)` - can we get a frequency histogram of <rmid, info> for a typical WAL?

That'll help us determine how much needs to be implemented in rust and where we can defer to redo code running elsewhere?
